File: kotlin-project/chain.kt
class Calculator {
    private var result: Double = 0.0

    fun add(value: Double): Calculator {
        result += value
        return this // 자기 자신 반환
    }

    fun subtract(value: Double): Calculator {
        result -= value
        return this // 자기 자신 반환
    }

    fun multiply(value: Double): Calculator {
        result *= value
        return this // 자기 자신 반환
    }

    fun divide(value: Double): Calculator {
        if (value != 0.0) {
            result /= value
        }
        return this // 자기 자신 반환
    }

    fun getResult(): Double {
        return result
    }
}

fun main() {
    val calc = Calculator()
    val result = calc.add(5.0)
        .multiply(2.0)
        .subtract(3.0)
        .divide(2.0)
        .getResult()
    println(result) // 출력: 4.0
}

File: cpp-project/chain.cpp
#include <iostream>

class Calculator {
private:
    double result;

public:
    Calculator() : result(0) {}

    Calculator& add(double value) {
        result += value;
        return *this; // 자기 자신 반환
    }

    Calculator& subtract(double value) {
        result -= value;
        return *this; // 자기 자신 반환
    }

    Calculator& multiply(double value) {
        result *= value;
        return *this; // 자기 자신 반환
    }

    Calculator& divide(double value) {
        if (value != 0) {
            result /= value;
        }
        return *this; // 자기 자신 반환
    }

    double getResult() const {
        return result;
    }
};

int main() {
    Calculator calc;
    double result = calc.add(5)
                        .multiply(2)
                        .subtract(3)
                        .divide(2)
                        .getResult();

    std::cout << result << std::endl; // 출력: 4
    return 0;
}